drivers/nec/display.cr

```diff
@@ -0,0 +1,299 @@
+    else
+      logger.debug { "-- NEC LCD, requested to power off" }
+      # 4 = Power Off
+      data = MsgType::Command.build(Command::SetPower, 4)
+      send(data, name: "power", delay: 10.seconds, timeout: 10.seconds)
+    end
+  end
+
+  def power?(**options) : Bool
+    data = MsgType::Command.build(Command::PowerQuery)
+    send(data, **options, name: "power?").get
+    self[:power].as_bool
+  end
+
+  def switch_to(input : Input)
+    logger.debug { "-- NEC LCD, requested to switch to: #{input}" }
+    data = MsgType::SetParameter.build(Command::VideoInput, input.value)
+    send(data, name: "input", delay: 6.seconds)
+  end
+
+  enum Audio
+    Audio1      = 1
+    Audio2      = 2
+    Audio3      = 3
+    Hdmi        = 4
+    Tv          = 6
+    DisplayPort = 7
+  end
+
+  def switch_audio(input : Audio)
+    logger.debug { "-- NEC LCD, requested to switch audio to: #{input}" }
+    data = MsgType::SetParameter.build(Command::AudioInput, input.value)
+    send(data, name: "audio")
+  end
+
+  def auto_adjust
+    data = MsgType::SetParameter.build(Command::AutoSetup, 1)
+    send(data, name: "auto_adjust")
+  end
+
+  def brightness(val : Int32)
+    data = MsgType::SetParameter.build(Command::BrightnessStatus, val.clamp(0, 100))
+    send(data, name: "brightness")
+    send(MsgType::Command.build(Command::Save), name: "brightness_save") # Save the settings
+  end
+
+  def contrast(val : Int32)
+    data = MsgType::SetParameter.build(Command::ContrastStatus, val.clamp(0, 100))
+    send(data, name: "contrast")
+    send(MsgType::Command.build(Command::Save), name: "contrast_save") # Save the settings
+  end
+
+  def volume(val : Int32)
+    data = MsgType::SetParameter.build(Command::VolumeStatus, val.clamp(0, 100))
+    send(data, name: "volume")
+    send(MsgType::Command.build(Command::Save), name: "volume_save") # Save the settings
+  end
+
+  def mute_audio(state : Bool = true, index : Int32 | String = 0)
+    logger.debug { "requested to update mute to #{state}" }
+    data = MsgType::SetParameter.build(Command::MuteStatus, state ? 1 : 0)
+    send(data, name: "mute_audio")
+  end
+
+  def do_poll
+    current_power = power?(priority: 0)
+    logger.debug { "Polling, power = #{current_power}" }
+
+    if current_power
+      mute_status
+      volume_status
+      video_input
+      audio_input
+    end
+  end
+
+  # LCD Response code
+  def received(data, task)
+    logger.debug { "task is #{task.try &.name}" }
+
+    ascii_string = String.new(data)
+    # Check for valid response
+    if !check_checksum(data)
+      return task.try &.retry("-- NEC LCD, invalid response was: #{ascii_string}")
+    end
+
+    logger.debug { "NEC LCD responded with ascii_string #{ascii_string}" }
+
+    command = MsgType.from_value(data[4])
+
+    case command # Check the MsgType (B, D or F)
+    when .command_reply?
+      # Power on and off
+      if ascii_string[10..15] == "C203D6" # Means power comamnd
```

Comment:
With some of the restructuring, these magic values can now likely be replaced by what's in the `Command` enum.

Reply:
I've made this commit but this still remains as the power query response is an annoyingly unique case.

https://github.com/PlaceOS/drivers/pull/50/commits/5d61fd20eddf243f99cc26bbe653a31b5181ee0f#diff-7c4dd3c6e0b44d710bc595343e7b26c65e527df44935f29f97927f6bd22c0904R153-R155

![image](https://user-images.githubusercontent.com/18463731/98016308-ebab7a00-1e51-11eb-9da4-e877468810e3.png)